src/metadataGeneration/tsoa.ts

```diff
@@ -95,48 +95,121 @@ export namespace Tsoa {
+   */
+  export type MetaType = PrimitiveMetaType | ObjectNoPropsMeta | EnumMeta | ArrayMetaType | DateTimeMeta | DateMeta | BinaryMeta | BufferMeta | ByteMeta | AnyMetaType | RefEnumMetaType | RefObjectMetaType | NestedObjectMetaType | UnionMetaType | IntersectionMetaType;
+
+  export interface StringMeta extends MetaTypeMinimal {
+      dataType: 'string'
+  }
+
+  export interface BooleanMeta extends MetaTypeMinimal {
+    dataType: 'boolean'
   }
 
-  export interface EnumerateType extends Type {
+  /**
+   * This is the type that occurs when a developer writes `const foo: object = {}` since it can no longer have any properties added to it.
+   */
+  export interface ObjectNoPropsMeta extends MetaTypeMinimal {
+    dataType: 'object'
+  }
+
+  export interface DoubleMeta extends MetaTypeMinimal {
+    dataType: 'double'
+  }
+
+  export interface FloatMeta extends MetaTypeMinimal {
+    dataType: 'float'
+  }
+
+  export interface IntegerMeta extends MetaTypeMinimal {
+    dataType: 'integer'
+  }
+
+  export interface LongMeta extends MetaTypeMinimal {
+    dataType: 'long'
+  }
+
+  export interface EnumMeta extends MetaTypeMinimal {
     dataType: 'enum';
     enums: string[];
   }
 
-  export interface ObjectLiteralType extends Type {
+  export interface ArrayMetaType extends MetaTypeMinimal {
+    dataType: 'array';
+
+    elementType: MetaType;
+  }
+
+  export interface DateMeta extends MetaTypeMinimal {
+    dataType: 'date'
+  }
+
+  export interface DateTimeMeta extends MetaTypeMinimal {
+    dataType: 'datetime'
+  }
+
+  export interface BinaryMeta extends MetaTypeMinimal {
+    dataType: 'binary'
+  }
+
+  export interface BufferMeta extends MetaTypeMinimal {
+    dataType: 'buffer'
+  }
+
+  export interface ByteMeta extends MetaTypeMinimal {
+    dataType: 'byte'
+  }
+
+  export interface VoidMeta extends MetaTypeMinimal {
+    dataType: 'void'
+  }
+
+  export interface AnyMetaType extends MetaTypeMinimal {
+    dataType: 'any'
+  }
+
+  export interface NestedObjectMetaType extends MetaTypeMinimal {
     dataType: 'nestedObjectLiteral';
     properties: Property[];
-    additionalProperties?: Type;
+    additionalProperties?: MetaType;
   }
 
-  export interface ArrayType extends Type {
-    dataType: 'array';
-    elementType: Type;
+  export interface RefEnumMetaType extends ReferenceTypeMinimal {
+    dataType: 'refEnum';
+    enums: string[] | number[];
   }
 
-  export interface ReferenceType extends Type, RefTypeMinimal {
+  export interface RefObjectMetaType extends ReferenceTypeMinimal {
+    dataType: 'refObject';
+    properties: Property[];
+    additionalProperties?: MetaType;
+  }
+
+  export type ReferenceType = RefEnumMetaType | RefObjectMetaType;
+
+  export interface ReferenceTypeMinimal extends MetaTypeMinimal {
     description?: string;
     dataType: RefTypeLiteral;
     refName: string;
-    properties?: Property[];
-    additionalProperties?: Type;
-    enums?: string[] | number[];
```

Comment:
@WoH these three lines above are probably the only part of this PR that can clarify the possible value that this PR provides. Right now it looks like a `ReferenceType` can have all of these properties or none. Which means that the compiler can’t help us enforce that we give the right properties to the right type.

Reply:
I certainly [agree](https://github.com/lukeautry/tsoa/pull/485/commits/a91dae8509ebed5cdae62acb14e824d1566d4db8) that this provides value ;)

Reply:
Okay yea, so 50% (or more) of the value here has already been achieved in your other PR. That’s more reason to close this.

Reply:
I disagree. The value does not only lie in fewer bugs/better checking of the existing code but a better DX for contributors. Let's not underestimate that